sdk/nodejs/appservice/zMixins.ts

```diff
@@ -711,3 +711,77 @@ export function getResourceGroupNameAndLocation(
     const getResult = resourceGroupName.apply(n => core.getResourceGroup({ name: n }));
     return { resourceGroupName, location: getResult.location };
 }
+
+/**
+ * Keys associated with a Function App.
+ */
+export interface FunctionHostKeys {
+    /** Master key. */
+    masterKey: string;
+    /** A dictionary of system keys, e.g. for Durable Functions or Event Grid. */
+    systemKeys: { [key: string]: string };
+    /** Default function keys. */
+    functionKeys: FunctionKeys;
+}
+
+/**
+ * Keys associated with a single Function.
+ */
+export interface FunctionKeys {
+    default: string;
+    [key: string]: string;
+}
+
+declare module "./functionApp" {
+    interface FunctionApp {
+        /**
+         * Retrieve the keys associated with the Function App.
+         */
+        getHostKeys(): pulumi.Output<FunctionHostKeys>;
+
+        /**
+         * Retrieve the keys associated with the given Function.
+         */
+        getFunctionKeys(functionName: pulumi.Input<string>): pulumi.Output<FunctionKeys>;
+    }
+}
+
+FunctionApp.prototype.getHostKeys = function(this: FunctionApp) {
+    return this.id.apply(async id => {
+        const credentials = await core.getServiceClientCredentials();
+        const client = new AzureServiceClient(credentials);
+        const url = `https://management.azure.com${id}/host/default/listkeys?api-version=2018-02-01`;
+
+        const response = await client.sendRequest({ method: "POST", url });
+        if (response.status >= 400) {
+            throw new Error(`Failed to retrieve the host keys: ${response.bodyAsText}`);
+        }
+
+        const body = response.parsedBody;
+        if (body.masterKey === undefined || body.systemKeys === undefined || body.functionKeys === undefined) {
+            throw new Error(`Wrong shape of the host keys response: ${response.bodyAsText}`);
+        }
+
+        return body as FunctionHostKeys;
+    });
+};
+
+FunctionApp.prototype.getFunctionKeys = function(this: FunctionApp, functionName) {
+    return pulumi.all([this.id, functionName]).apply(async ([id, name]) => {

```

Comment:
i'm feeling dumb, but i cannot understand what this functionName is for.  is it just to ensure that hte function has been created first before getFunctionKeys has been called?  it seems weird...

Reply:
Its used below in line 773 in the API call to the ARM Api to identify the function inside of the function app.

Reply:
Each function has its own set of keys in addition to the host keys, so we request URL contains the name as @ChristianEder kindly noticed

Reply:
oh... doh.  I see now.  i was looking for functionName, and not 'name'.  In general we actually *like* reusing the names (i.e. `pulumi.all([this.id, functionName]).apply(async ([id, functionName])`.  It helps indicate they represent the same thing, and it makes sure that the outer variable is blocked from being used in the inner scope.  I was looking for that pattern and that's what threw me off.  This is all good, thanks for clarifying!

Reply:
That makes sense, I'll rename and pick up this pattern.